### ISSUE TITLE
Plugin Widget Question

### ISSUE THREAD

**Issue by dcmcshan** [Syzygyx/DroneWidgets#30]
Possible to setup an architecture where we can compile the widgets separately without recompiling the whole executable?  So we'd have a base version of the QGC fork and just compile up widgets as we go along.

**Comment by BojanKverh** [comment]
Yes, it is possible, although I have never done this before. Which widgets would you like to compile separately? Also, some code is already integrated in the original qgroundcontrol files, like MainWindow.cc and QGCMapWidget.cc, so that will be much more difficult to take out and compile as a separate plugin.

**Comment by dcmcshan** [comment]
Just thinking ahead and wanting an easily extensible architecture for customers.  In general, we would have a main product and then be able to send them plugins for new features.  For instance, if we wanted to package this up for a different customer, would be great to have a "branding" package with the icons and splash as well as widgets specifically for their aircraft.  Then we wouldn't have to maintain an executable for each customer.

**Comment by BojanKverh** [comment]
I understand, I think it is a good approach. I will keep that in mind. But things like the GeoFence zones, adding waypoints to a mission by double clicking and similar, would probably be universal to all customers, right?

**Comment by BojanKverh** [comment]
With the extensive help from (very good) manuals found on Internet, I managed to add functionality to QGC through a plugin loader. Now loading the splash screen is done by ReferenceTechnologiesPlugin.so library (remove this file from plugins/lib folder and the splash screen pixmap will not appear). Now we could move gauges and maybe some other stuff also into plugins.

**Comment by dcmcshan** [comment]
Cool.  I’ll figure out how we want to bundle things, but in general, maybe it makes sense to just put each widget in a plugin?

I’ve also moved syzygyx/qgroundcontrol to the syzygyx/zyg repositiory.  It will be private.  Let me know if you can use it.  

I think what I actually want to do is to have our qgroundcontrol repository just have the plugin loader changes, and all the widgets will live in the /zyg directory.  This way, we can keep qgroundcontrol updated as the other repo evolves.

**Comment by BojanKverh** [comment]
I can access the zyg repo, except that the latest commit from today is missing (the one implementing plugins).

Regarding the plugin and qgroundcontrol repo separation; the idea is good and most of the widgets can be moved into the plugin repo. But I don't know about keeping qgroundcontrol updated: the problem is that some changes had to be made in the original qgroundcontrol code in order to accommodate our new functionalities and trying to keep the QGC code up to date will delete our changes, unless both branches can be merged somehow retaining all changes? Even if I try to move all my code out of QGC repo, at least the code that loads a plugin and generates missing functionality still has to stay there, otherwise QGC will just be the old QGC. Even if I would try to put the added code into classes, which inherit the original QGC ones, there is still high probability that somehing in the old QGC code would have to be changed (like methods made virtual, etc)

**Comment by dcmcshan** [comment]
Right, there will have to be our plugin changes in the forked repo.  And we’ll have to merge from time to time, but that is what it is.

I can’t easily get your last commit over to the zyg repo, but sounds like what we want to do is to leave those changes in the qgroundcontrol directory anyways, and just put plugins in the zyg repo.

**Comment by BojanKverh** [comment]
Yes, interfaces and plugins folder I think

**Comment by BojanKverh** [comment]
OK, I transferred the plugins folder, interfaces I think it's better to leave it in qgroundcontrol

**Comment by dcmcshan** [comment]
Okay, you can clean up the zyg repository.  It doesn't need all of qgroundcontrol in it

**Comment by BojanKverh** [comment]
Today I will transfer as many widgets as possible into plugins. For other functionalities that still need to be done in the future, please feel free to open new issue threads.

**Comment by BojanKverh** [comment]
All the flight indicators are now available as plugins. I think Mission planning, Vehicle widget and FlightGear video feed still remain to be ported.

**Comment by BojanKverh** [comment]
Moved Mission planning, vehicle widget and FlightGear video feed widget into the plugin library. Please let me know, if there is anything else we should move there.

**Comment by dcmcshan** [comment]
Should the zyg plugin library have the full qgrouncontrol app?  Does it need it to build?  If not, maybe we should clean it so that it just builds the plugins?

**Comment by BojanKverh** [comment]
zyg plugin only needs interfaces folder from qgroundcontrol. I already thought about moving that into zyg. However, one has to depend on the other or the other way around, they cannot be two completely different entities, because they use the same interfaces. If you put qgroundcontrol folder and zyg folder into the same place (folder), both should compile well.

**Comment by BojanKverh** [comment]
Were you able to compile and test the app and plugins then?

**Comment by dcmcshan** [comment]
Going to check it out on Windows today!

What do I build for the plugins?  Is there a Qt Creator file just for plugins?

**Comment by BojanKverh** [comment]
It is in zyg/plugins/ReferenceTechnologies/ReferenceTechnologies.pro
Load that file with Qt creator and then build.

**Comment by BojanKverh** [comment]
Did you manage to build the app and plugins? There is also the installation package prepared for Windows on DropBox.

**Comment by BojanKverh** [comment]
Hi Dan, how is it going?
I did a couple of corrections today on qgroundcontrol, that I haven't noticed so far. The plugin path on Linux should now be ok and also the plugin widgets are visible immediately after loading at startup.
